src/Subscriber.ts

```diff
@@ -83,18 +83,33 @@ export class Subscriber<T> extends Subscription implements Observer<T> {
 
 class SafeSubscriber<T> extends Subscriber<T> {
 
-  constructor(next?: (x?: T) => void,
+  constructor(private parent?: Subscriber<T>,
+              observerOrNext?: Observer<T> | ((value: T) => void),
               error?: (e?: any) => void,
               complete?: () => void) {
     super();
-    this._next = (typeof next === 'function') && tryOrThrowError(next) || null;
-    this._error = (typeof error === 'function') && tryOrThrowError(error) || throwError;
-    this._complete = (typeof complete === 'function') && tryOrThrowError(complete) || null;
+
+    let next: ((value: T) => void);
+    let context: any = this;
+
+    if (isFunction(observerOrNext)) {
+      next = (<((value: T) => void)> observerOrNext);
+    } else if (observerOrNext) {
+      context = observerOrNext;
+      next = (<Observer<T>> observerOrNext).next;
+      error = (<Observer<T>> observerOrNext).error;
+      complete = (<Observer<T>> observerOrNext).complete;
+    }
+
+    this._next = isFunction(next) && tryOrHandleMessageError(next, context) || null;
+    this._error = tryOrHandleMessageError(isFunction(error) && error || throwError, context);
+    this._complete = isFunction(complete) && tryOrHandleMessageError(complete, context) || null;
   }
 
-  next(value?: T): void {
+  next(value: T): void {
     if (!this.isStopped && this._next) {
       this._next(value);
+      this._reportMessageError(false);
```

Comment:
since `_reportMessageError` doesn't do anything unless `syncErrorThrown` is `true`, shouldn't we wrap this in a conditional?

```js
if (this.syncErrorThrown) {
  this._reportMessageError(false);
}
```